### ISSUE TITLE
fail to decode events of custom chain

### ISSUE THREAD

**Issue by sergeibadass** [itering/scale.go#11]
Hi may i ask what is the CRAB2  here in the event_test.go?

``` go
func TestCrabEventsDecoder(t *testing.T) {
	m := scalecodec.MetadataDecoder{}
	m.Init(utiles.HexToBytes(Crab2))
	_ = m.Process()
	c, err := ioutil.ReadFile(fmt.Sprintf("%s.json", "network/crab"))
	if err != nil {
		panic(err)
	}
	types.RegCustomTypes(source.LoadTypeRegistry(c))
	e := scalecodec.EventsDecoder{}
	option := types.ScaleDecoderOption{Metadata: &m.Metadata}

	eventRaw := "0x180000000000000080e36a090000000002000000010000000000000000000000000002000000020000000e022ac9219ace40f5846ed675dded4e25a1997da7eabdea2f78597a71d6f38031487089d481874e06bd4026807fd0464f5c7a1c691c21237ed9b912ac0443a2bc2200ca9a3b000000000000000000000000000002000000150600b4c4040000000000000000000000000000020000000e04b4f7f03bebc56ebe96bc52ea5ed3159d45a0ce3a8d7f082983c33ef133274747002d31010000000000000000000000000000020000000000401b5f1300000000000000"
	e.Init(types.ScaleBytes{Data: utiles.HexToBytes(eventRaw)}, &option)
	e.Process()
	b, _ := json.Marshal(e.Value)
	fmt.Println(string(b))
}
```

and how to generate CRAB2 for note-template chains with custom event types? Thanks.

**Comment by sergeibadass** [comment]
I managed to query the metadata from the node and import the hex of metadata like Crab2 and Kusama. Then i run into following errors in the test code. 

I just the human-readable form of the metadata. the event type is complete

```go
/usr/lib/go-1.13/bin/go tool test2json -t /tmp/___TestEventsDecoder_in_github_com_itering_scale_go -test.v -test.run ^\QTestSpannerEventsDecoder\E$
=== RUN   TestSpannerEventsDecoder
--- FAIL: TestSpannerEventsDecoder (0.19s)
panic: Not find Extrinsic Lookup 0020, please check metadata info [recovered]
	panic: Not find Extrinsic Lookup 0020, please check metadata info
```

**Comment by sergeibadass** [comment]
@freehere107 @yrong after some serious digging through the code, I think your current event decoding implementation does not support the latest version of substrate. 

Specifically, it fails to decode the basic system event DispatchInfo. I matched the output of your GO SCALE codec results with that of Polkadot.{js} bit-by-bit and no luck. Substrate chain spec version 259

It can get the weight right but then the call_class wrong. and thus everything thereafter. For example, for event hex 
0x0800000000000000482d7c0900000000020000000100000000000000000000000000020000 which are 
``` bash
system:ExtrinsicSuccess:: (phase={"ApplyExtrinsic":0})
    DispatchInfo: {"weight":159133000,"class":"Mandatory","paysFee":"Yes"}
system:ExtrinsicSuccess:: (phase={"ApplyExtrinsic":1})
    DispatchInfo: {"weight":0,"class":"Mandatory","paysFee":"Yes"}
```
failed on SCALE.go. 

Any good pointer to solve this problem?

**Comment by freehere107** [comment]
Thank you feedback，I will relay you tomorrow

**Comment by freehere107** [comment]
hey, Crab2 is crab network runtime metadata raw, this used to decode Extrinsic/Event.

If decode fails, you can check  the next few steps
1. runtime metadata first
2. type JSON, The  structure of each network is different, you need to configure the appropriate data structure yourself, like this https://github.com/itering/scale.go/tree/master/network

**Comment by freehere107** [comment]
test case templete

```
func TestCanvasExtrinsicDecoder(t *testing.T) {
	m := scalecodec.MetadataDecoder{}
	m.Init(utiles.HexToBytes(``)) // Todo: rpc state_getMetadata
	_ = m.Process()
	c, err := ioutil.ReadFile(fmt.Sprintf("%s.json", "network/custom")) // Todo: your custom network 
	if err != nil {
		panic(err)
	}
	types.RegCustomTypes(source.LoadTypeRegistry(c))

	e := scalecodec.ExtrinsicDecoder{}
	option := types.ScaleDecoderOption{Metadata: &m.Metadata}
	extrinsicRaw := ""  // todo extrinsic raw data
	e.Init(types.ScaleBytes{Data: utiles.HexToBytes(extrinsicRaw)}, &option)
	e.Process()
	fmt.Println(utiles.ToString(e.Value))
}
```

**Comment by sergeibadass** [comment]
@freehere107 thanks for your reply.

I have configured the network types.json file. My complete status is
(1) Metadata test passed
(2) Extrinsic decoding test passed.
(3) Subscan can [index my blocks and extrinsics except events
(4) Event decoding FAILED

```go
func TestCustomExtrinsicDecoder(t *testing.T) {
	m := scalecodec.MetadataDecoder{}
	m.Init(utiles.HexToBytes(customMeta))
	_ = m.Process()
	c, err := ioutil.ReadFile(fmt.Sprintf("%s.json", "network/custom"))
	if err != nil {
		panic(err)
	}
	types.RegCustomTypes(source.LoadTypeRegistry(c))
	types.RegCustomTypes(source.LoadTypeRegistry([]byte(`{"Address": "GenericAddress","LookupSource": "GenericAddress"}`)))
	e := scalecodec.ExtrinsicDecoder{}
	option := types.ScaleDecoderOption{Metadata: &m.Metadata}
	extrinsicRaw := "0x2d0284ffd43593c715fdd31c61141abd04a99fd6822c8558854ccde39a5684e7a56da27d01e0c80f699a2a7c596153795c41163b64e8a7fff39fec8ab1871b8f91a5fe670d6e9be795c4e7e3867a1f1d499be2d5623804e6e956db39771a3b2815fc27f28b06013c002000000000a0724e18090000000000000000000000e87648170000000000000000000000"
	e.Init(types.ScaleBytes{Data: utiles.HexToBytes(extrinsicRaw)}, &option)
	e.Process()
	fmt.Println(utiles.ToString(e.Value))
}

```

extracted from this file. (rename to metadata.json)
[metadata.log](https://github.com/itering/scale.go/files/5923010/metadata.log)


download this file and rename it to custom.json (github issue does not allow me to update json)
[custom.log](https://github.com/itering/scale.go/files/5923003/custom.log)

In short, everything works except events. Can you help? @freehere107

**Comment by freehere107** [comment]
@sergeibadass  Can you give me your  event raw data

**Comment by sergeibadass** [comment]
nevermind now i fixed the problem. the base decoder Weight is u32. let me help u close this issue.
